### DISCUSSION TITLE
Transfer-Encoding:chunked导致websocket handshake hang住

### DISCUSSION THREAD

**Discussion by haniceboy** [bfenetworks/bfe#827] [General]
大家好。我们在自己流量架构中使用了开源的bfe（0.12）进行流量转发。我们现在遇到了问题，
    
- 问题现象：当client使用websocket连接的时候（例如：ws://demo.baidu.com/ws）， 会一直hang在websocket connecting,
-  原因：通过抓包和看bfe的源码，我们发现bfe收到envoy返回的http code 101的response中包含了header Transfer-Encoding:chunked所导致。bfe会一直等待额外的body传输，但是101 response是不含包体的。而这个额外的Transfer-Encoding:chunked是envoy给添加的。
- 我们的流量架构：百度云blb->开源bfe->envoy(istio))->服务
- 我们通过fix1解决了这个问题，其详情如下：
![1](https://user-images.githubusercontent.com/5663566/133059661-0ee97a17-287d-4ba6-9507-90a95cb4901e.jpg)

fix1只解决了我们大部分的websocket连接,我们还有如下的问题没有解决，我们发现bfe单独代理websocket协议时也会添加Transfer-Encoding:chunked，这导致其它bfe转发过来的websocket流量，还是会有相同的问题。我们已经找到了在101 response 中添加Transfer-Encoding:chunked的位置（请参加下图），
![2](https://user-images.githubusercontent.com/5663566/133060101-c3366f2e-1ef0-406c-b679-a3834b11398d.jpg)

但是不太理解其逻辑。我们只能添加一个临时fix2(把隶属于websocket 101 response中的Transfer-Encoding:chunked删掉，以便兼容我们的bfe接在其它bfe后面的情况。(fix2详情如下图)
![3](https://user-images.githubusercontent.com/5663566/133060155-db973297-d456-477c-baee-cfbeee3fda81.png)

我们想咨询的问题是：
1.bfe在101 response 中添加Transfer-Encoding:chunked的逻辑是什么？是否为标准协议的一部分（因为envoy也有相同的行为）？
2.bfe是否需要fix1的修改？因为如果所有的bfe有类似fix1的处理，我们就不需要添加fix2的修改。

**Comment by iyangsj** [comment]
感谢您反馈的兼容性问题及详细情况
 1.对于响应头部未包含Content-Length或无法确定长度的场景，目前的程序在转发时会添加chunked头部处理
 2.我们将评估您的修复建议，并在下个版本发布最终解决方案
